parlai/core/dialog_teacher.py

```diff
@@ -96,37 +81,21 @@ def label_candidates(self):
         """
         return None
 
-    def observe(self, observation):
-        """Process observation for metrics."""
-        if self.lastY is not None:
-            self.metrics.update(observation, self.lastY)
-            self.lastY = None
-        return observation
-
     def next_example(self):
         num_eps = self.data.num_episodes()
         if not self.stream:
             if self.episode_done:
-                if self.random:
-                    # select random episode
-                    self.episode_idx = random.randrange(num_eps)
-                else:
-                    # select next episode
-                    self.episode_idx = (self.episode_idx + self.step_size) % num_eps
+                self.episode_idx, self.epochDone = self.next_episode_idx(num_eps=num_eps)
                 self.entry_idx = 0
             else:
                 self.entry_idx += 1
             action, epoch_done = self.data.get(self.episode_idx, self.entry_idx)
         else:
             action, epoch_done = self.data.get()
-
-        if self.random:
-            epoch_done = False
-        elif (self.episode_idx + self.step_size >= num_eps and
-                action['episode_done'] and not self.stream):
+        if not self.random and action['episode_done'] and not self.stream:
             # this is used for ordered data to check whether there's more data
             epoch_done = True
-        return action, epoch_done
+        return action, self.epochDone or epoch_done
```

Comment:
is this right? maybe the `self.epochDone =` should be `epoch_done =`

Reply:
Yeah you're right, that's what it should be